Add mWh as unit of measurement for Matter energy sensors

agners opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The Matter energy measurement exposes energy consumption in milliwatts per hour. Add this as a unit of measurement so we can store the data in the Matter native unit.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/36310 https://github.com/home-assistant/developers.home-assistant/pull/2493

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

agners committed:
Add mWh as unit of measurement for Matter energy sensors

The Matter energy measurement exposes energy consumption in milliwatts
per hour. Add this as a unit of measurement so we can store the
data in the Matter native unit.

epenet reviewed (commented):
I think this needs corresponding docs + dev docs

We also have the units listed in the docstrings for SensorDeviceClass.ENERGY / SensorDeviceClass.ENERGY_STORAGE and corresponding NumberDeviceClass.ENERGY / NumberDeviceClass.ENERGY_STORAGE but these might need to be realigned correctly as the four do not seem to match anymore

agners committed:
Fix sensor tests

agners committed:
Update DeviceClass docstrings

epenet commented:
> The integration docs don't really mention Energy measurement units, what do you expect me to update there?

I was talking about the number/sensor integration docs

https://github.com/home-assistant/home-assistant.io/blob/f4c9bc02c327af4f16fcb9417493822bec141837/source/_integrations/sensor.markdown?plain=1#L65-L66

https://github.com/home-assistant/home-assistant.io/blob/f4c9bc02c327af4f16fcb9417493822bec141837/source/_integrations/number.markdown?plain=1#L54-L55

agners committed:
Fix config flow tests

agners commented on tests/components/template/test_config_flow.py:
```diff
@@ -804,7 +804,7 @@ async def test_config_flow_preview(
                 "unit_of_measurement": (
                     "'None' is not a valid unit for device class 'energy'; "
-                    "expected one of 'cal', 'Gcal', 'GJ', 'GWh', 'J', 'kcal', 'kJ', 'kWh', 'Mcal', 'MJ', 'MWh', 'TWh', 'Wh'"
+                    "expected one of 'cal', 'Gcal', 'GJ', 'GWh', 'J', 'kcal', 'kJ', 'kWh', 'Mcal', 'MJ', 'mWh', 'MWh', 'TWh', 'Wh'"
```
Hm, this test did pass locally.

Now running it multiple times, it seems the order is random here.

The reason is that we sort the units with `str.casefold`:

https://github.com/home-assistant/core/blob/4e5df1d854a4926bedc474aa9c16ff70057ebbdc/homeassistant/components/template/config_flow.py#L240

Which made me realize that the new unit `mWh` is actually equal to `MWh` when used case insensitivity. Is that acceptable?

From what I understand this is really the [SI/metric prefixes](https://en.wikipedia.org/wiki/Metric_prefix). I guess we might run into this in other places too, but currently I don't see any precedence :thinking: 

I try changing the above sort and see how things are going.

agners committed:
Make sure template validation errors are consistent

Template validation errors were not consistent when two unit of
measurement had the same (case-insensitive) name (e.g. mWh and MWh).
This makes sure the error is always consistent.

abmantis commented on homeassistant/util/unit_conversion.py:
```diff
@@ -266,6 +266,7 @@ class EnergyConverter(BaseUnitConverter):
         UnitOfEnergy.KILO_JOULE: _WH_TO_J,
         UnitOfEnergy.MEGA_JOULE: _WH_TO_J / 1e3,
         UnitOfEnergy.GIGA_JOULE: _WH_TO_J / 1e6,
+        UnitOfEnergy.MILLIWATT_HOUR: 1e6,
```
```suggestion
        UnitOfEnergy.MILLI_WATT_HOUR: 1e6,
```
to keep it consistent with the other energy units (which are not consistent with other units :upside_down_face: which have no underscore)

frenck reviewed (approved):
Thanks, @agners 👍 

../Frenck
